src/ingest-contact-loaders/csv-upload/index.js

```diff
@@ -0,0 +1,136 @@
+  /// Sometimes credentials need to be setup, etc.
+  /// A second key expiresSeconds: should be how often this needs to be checked
+  /// If this is instantaneous, you can have it be 0 (i.e. always), but if it takes time
+  /// to e.g. verify credentials or test server availability,
+  /// then it's better to allow the result to be cached
+  return {
+    result: true,
+    expiresSeconds: 0
+  }
+}
+
+export function addServerEndpoints(expressApp) {
+  /// If you need to create API endpoints for server-to-server communication
+  /// this is where you would run e.g. app.post(....)
+  /// Be mindful of security and make sure there's
+  /// This is NOT where or how the client send or receive contact data
+  return;
+}
+
+export function clientChoiceDataCacheKey(organization, campaign, user) {
+  /// returns a string to cache getClientChoiceData -- include items that relate to cacheability
+  //return `${organization.id}-${campaign.id}`;
+  return "";
+}
+
+export async function getClientChoiceData(organization, campaign, user, loaders) {
+  /// data to be sent to the admin client to present options to the component or similar
+  /// The react-component will be sent this data as a property
+  /// return a json object which will be cached for expiresSeconds long
+  /// `data` should be a single string -- it can be JSON which you can parse in the client component
+  return {
+    data: "",
+    expiresSeconds: 0
+  };
+}
+
+export async function processContactLoad(job, maxContacts) {
+  /// trigger processing -- this will likely be the most important part
+  /// you should load contacts into the contact table with the job.campaign_id
+  /// Since this might just *begin* the processing and other work might
+  /// need to be completed asynchronously after this is completed (e.g. to distribute loads)
+  /// AFTER true contact-load completion, this (or another function) MUST call
+  /// src/workers/jobs.js::completeContactLoad(job)
+  ///   The async function completeContactLoad(job) will delete opt-outs, delete duplicate cells, clear/update caching, etc.
+  /// Basic responsibilities:
+  /// 1. delete previous campaign contacts on a previous choice/upload
+  /// 2. set campaign_contact.campaign_id = job.campaign_id on all uploaded contacts
+  /// 3. set campaign_contact.message_status = "needsMessage" on all uploaded contacts
+  /// 4. Ensure that campaign_contact.cell is in the standard phone format "+15551234567"
+  ///    -- do NOT trust your backend to ensure this
+  /// Things to consider in your implementation:
+  /// * Batching
+  /// * Error handling
+  /// * "Request of Doom" scenarios -- queries or jobs too big to complete
+  const campaignId = job.campaign_id;
+  const jobMessages = [];
+
+  await r
+    .knex("campaign_contact")
+    .where("campaign_id", campaignId)
+    .delete();
+
+  // TODO
+  let contacts;
+  if (job.payload[0] === '{') {
+    contacts = JSON.parse(job.payload).contacts;
```

Comment:
if `JSON.parse` fails, it will return an error. Should this be a `try catch` block? Could that `try catch` even replace the kind of hacky if statement checking for `{`? Not sure how often we've seen problems with `JSON.parse` failing, but it'd be nice to use the `catch` to surface a user facing message

Reply:
If JSON.parse can fail, then the tests will fail.  It should be impossible.  Also note that this is partially guarded already because we confirm tahat the first char is a `{` (a good sign of JSON, anyway)-- I think it's ok to make this an assumption, and if it throws/errors, then the application error log will be visible.